src/backend/buffer.rs

```diff
@@ -118,39 +155,38 @@ where
     unsafe fn resolve_index_to_str_unchecked(&self, index: usize) -> &str {
         // SAFETY: The function is marked unsafe so that the caller guarantees
         //         that required invariants are checked.
-        let slice_len = unsafe { self.buffer.get_unchecked(index..) };
-        // SAFETY: The function is marked unsafe so that the caller guarantees
-        //         that required invariants are checked.
-        let (str_len, str_len_bytes) = unsafe { decode_var_usize_unchecked(slice_len) };
-        let start_str = index + str_len_bytes;
-        let str_bytes =
-            // SAFETY: The function is marked unsafe so that the caller guarantees
-            //         that required invariants are checked.
-            unsafe { self.buffer.get_unchecked(start_str..start_str + str_len) };
+        let len_bytes =
+            unsafe { self.buffer.get_unchecked(index - LEN_LOOKBEHIND..index) };
+        let len_bytes = <&[u8; LEN_LOOKBEHIND]>::try_from(len_bytes).unwrap();
+        let len = decode_len_backwards(len_bytes);
+        let str_bytes = unsafe { self.buffer.get_unchecked(index..index + len) };
```

Comment:
Would this fix https://github.com/Robbepop/string-interner/issues/49 ?

Reply:
No; the usize value can take any set of 8 bytes, some of which may fall outside the utf8 encoding.

One possibility is to make a more complex encoding for the usize, with 7 bits per byte, leaving the 8th bit unset. This would keep it within the ASCII range. This requires some encoding/decoding. The problem is that this encoding/decoding requires a shift and an or, per byte, which would defeat most of the performance gains in this PR.

On x86 machines with fast pext/pdep instructions (Intel Haswell and later; AMD Zen 3 and later) we can use `pext rax, 0x7f7f7f7f7f7f7f7f` for this decoding and `pdep rax, 0x7f7f7f7f7f7f7f7f` for the encoding. Sadly, older machines either don't have these instruction (Intel) or run them very slowly (Zen 1 and Zen 2).

The "bit-table indicating string start" approach would allow a fast (single-instruction) check, but at the cost of 12.5% space overhead. If we required strings to be aligned at 2 bytes, we could have a bit per 2 bytes (6.25% overhead) or a bit per 4 bytes (3.125% overhead), etc.

I'm not sure which approach is best. I'd probably go with a bit-table with 2 byte alignment as the least-worst compromise.